In [1]:
import tensorflow_hub as hub
import tensorflow_text as text

In [2]:
preprocess_url = "https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3"
encoder_url = "https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4"

In [3]:
bert_preprocess_model = hub.KerasLayer(preprocess_url)

In [14]:
text_test = ['It was a wonderfull movie ','I hate Fake people']
text_preprocessed = bert_preprocess_model(text_test)
text_preprocessed.keys()

dict_keys(['input_type_ids', 'input_mask', 'input_word_ids'])

In [15]:
text_preprocessed['input_mask']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      dtype=int32)>

In [16]:
text_preprocessed['input_type_ids']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      dtype=int32)>

In [17]:
text_preprocessed['input_word_ids']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[ 101, 2009, 2001, 1037, 6919, 2140, 3185,  102,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0],
       [ 101, 1045, 5223, 8275, 2111,  102,    0,    0,    0,    0,    0,
           0,

**101 --> CLS token**

**102 --> SEP token**

BERT uses CLS as a special token at the begining of each setence whereas SEP as a special token to
separate two sentences or end sinle sentece

In [18]:
bert_model = hub.KerasLayer(encoder_url)

In [19]:
bert_results = bert_model(text_preprocessed)

In [20]:
bert_results.keys()

dict_keys(['pooled_output', 'sequence_output', 'encoder_outputs', 'default'])

In [21]:
bert_results['sequence_output']

<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
array([[[ 0.03256144,  0.0162181 ,  0.1641414 , ..., -0.11029746,
          0.29263777,  0.13337998],
        [-0.22427258, -0.32069042, -0.05837841, ..., -0.33245036,
          0.9609185 ,  0.30647847],
        [-0.08641529, -0.45443314, -0.03351502, ..., -0.1476955 ,
          0.4753579 ,  0.02004288],
        ...,
        [ 0.08778356, -0.14280796,  0.77970195, ...,  0.12068477,
          0.18778831, -0.0344475 ],
        [ 0.00251873, -0.1656923 ,  0.74947053, ...,  0.13749151,
          0.15921041, -0.01454027],
        [-0.14639714, -0.05747539,  0.683909  , ..., -0.00141143,
         -0.01543484,  0.18695259]],

       [[ 0.33239853,  0.43038517, -0.13503098, ..., -0.03596561,
          0.16116676,  0.28474793],
        [ 0.42021552,  0.23694837, -0.8452966 , ...,  0.12228374,
          0.698787  ,  0.2714639 ],
        [ 0.7275566 ,  0.9529403 , -0.3052128 , ...,  0.03852374,
          0.3723142 ,  0.3102372 ],
        ...,

In [22]:
bert_results['pooled_output']

<tf.Tensor: shape=(2, 768), dtype=float32, numpy=
array([[-0.71888745, -0.13307711,  0.46135545, ...,  0.27829376,
        -0.5267889 ,  0.80517584],
       [-0.78882635, -0.222432  ,  0.55274355, ...,  0.48104456,
        -0.5025707 ,  0.8637377 ]], dtype=float32)>

In [23]:
len(bert_results['encoder_outputs'])

12

Since we are using BERT base model it has 12 encoder layers, that's why above the length of encoder_output is 12.


You can see below that last element of encoder_outputs is basically a sequence_output

In [24]:
bert_results['encoder_outputs'][-1] == bert_results['sequence_output']

<tf.Tensor: shape=(2, 128, 768), dtype=bool, numpy=
array([[[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],

       [[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]]])>

In [25]:
bert_results['encoder_outputs']

[<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
 array([[[ 1.68860108e-01,  5.45844734e-02, -5.93563169e-02, ...,
           1.09595463e-01,  8.13109875e-02,  3.56395617e-02],
         [-3.30172926e-01, -1.34215415e-01,  2.75766879e-01, ...,
           7.76617050e-01,  6.79530621e-01, -7.83272460e-02],
         [-7.01308966e-01, -4.05519217e-01,  6.57428920e-01, ...,
          -7.46893883e-03, -5.69414161e-02, -7.87657499e-02],
         ...,
         [-1.18952811e-01, -1.65546522e-01,  6.43001974e-01, ...,
           5.62752903e-01, -2.25746840e-01,  5.05190715e-02],
         [-2.00518072e-01, -1.83851019e-01,  5.38432240e-01, ...,
           5.76573372e-01, -1.33847162e-01, -3.43384743e-02],
         [-6.94517940e-02, -1.20781332e-01,  5.60043335e-01, ...,
           8.52476418e-01, -4.19101626e-01, -9.85443667e-02]],
 
        [[ 2.26572052e-01,  2.00698152e-04, -9.95762497e-02, ...,
           9.95486975e-04,  1.21075146e-01,  1.56923309e-02],
         [ 9.10385430e-01,  8.1